src/parser/common.rs

```diff
@@ -155,6 +155,43 @@ pub fn parse_gate_modifier<'a>(input: ParserInput<'a>) -> ParserResult<'a, GateM
     ))
 }
 
+/// Parse matrix used to define gate with `DEFGATE`.
+pub fn parse_matrix<'a>(input: ParserInput<'a>) -> ParserResult<'a, Vec<Vec<Expression>>> {
+    let (input, matrix) = separated_list0(
+        token!(NewLine),
+        preceded(
+            many1(token!(Indentation)),
+            separated_list0(token!(Comma), parse_expression),
+        ),
+    )(input)?;
+    Ok((input, matrix))
+}
+
+/// Parse permutation representation of a `DEFGATE` matrix.
+pub fn parse_permutation<'a>(input: ParserInput<'a>) -> ParserResult<'a, Vec<Vec<Expression>>> {
```

Comment:
Per the spec, this is a comma-separated list of integers, not expressions: https://quil-lang.github.io/ sect 4.2.3